Deploy to OpenShift command

datho7561 opened the pull request:
Uses the existing menu to select which Quarkus project to deploy.
Uses OpenShift Connector in order to deploy the Quarkus project to
OpenShift.
OpenShift Connector handles selecting/creating the application and
naming the component.

Closes #313

Signed-off-by: David Thompson <davthomp@redhat.com>

dgolovin commented on src/wizards/deployToOpenShift/deployToOpenShift.ts:
```diff
@@ -0,0 +1,51 @@
+  try {
+    const openShiftConnectorExtension: any = await installOpenShiftConnectorIfNeeded();
+    const quarkusProject: ProjectLabelInfo = await getQuarkusProject();
+    // openShiftConnectorExtension.Component.createFromFolder(quarkusProject.uri);
```
Considering we don't have API for that yet, I would suggest to use vscode 'commands' API. Here is the issue for extension to provide vscode command for that https://github.com/redhat-developer/vscode-openshift-tools/issues/1832.

fbricon reviewed (changes requested):
Didn't work for me:

<img width="553" alt="Screenshot 2020-11-13 at 10 40 15" src="https://user-images.githubusercontent.com/148698/99058240-36896800-259d-11eb-95bd-13b20d020048.png">

Maybe it's because I'm running vscode-quarkus from the Extension Development Host? The openshift extension does install though

fbricon commented on src/utils/openShiftConnectorUtils.ts:
```diff
@@ -0,0 +1,38 @@
+ * Resolves when the extension is installed
+ */
+export async function installOpenShiftConnector(): Promise<any> {
+  commands.executeCommand("workbench.extensions.installExtension", OPENSHIFT_CONNECTOR_EXTENSION_ID);
```
```typescript
export async function installOpenShiftConnector(): Promise<any> {
  return window.withProgress({ location: ProgressLocation.Notification, title: `Installing ${OPENSHIFT_CONNECTOR}...`}, progress => {
    return commands.executeCommand("workbench.extensions.installExtension", OPENSHIFT_CONNECTOR_EXTENSION_ID);
  }).then(() => {
    if (!!extensions.getExtension(OPENSHIFT_CONNECTOR_EXTENSION_ID)) {
      window.showInformationMessage(`Successfully installed ${OPENSHIFT_CONNECTOR}.`);
    } else {
      window.showErrorMessage(`Failed to install ${OPENSHIFT_CONNECTOR}.`);
    }
  });
```

fbricon replied:
@jrieken any idea why calling extensions.getExtension(...) after calling "workbench.extensions.installExtension" always fails?

fbricon replied:
Ok so my guess is we need to add a listener to extensions.onDidChange, once we detect the OPENSHIFT_CONNECTOR_EXTENSION_ID has been registered, then we can proceed with calling the openshift deployment command

datho7561 replied:
I improved the error reporting on failed extension installation. It properly waits for OpenShift Connector to finish installing. I also added a timeout for the installation handling code.

fbricon commented on src/wizards/deployToOpenShift/deployToOpenShift.ts:
```diff
@@ -0,0 +1,51 @@
+  try {
+    await installOpenShiftConnectorIfNeeded();
+    const quarkusProject: ProjectLabelInfo = await getQuarkusProject();
+    await commands.executeCommand("openshift.component.createFromRootWorkspaceFolder", quarkusProject.uri, "java-quarkus");
```
Once a cluster is already available (or else it [fails](https://github.com/redhat-developer/vscode-openshift-tools/issues/1849)), the command lets you create an [application](https://github.com/redhat-developer/vscode-openshift-tools/issues/1850) and a component, if needed.

Then it fails with:

<img width="539" alt="Screenshot 2020-11-13 at 18 03 23" src="https://user-images.githubusercontent.com/148698/99099398-ac5ff480-25da-11eb-978b-5e85b4513a03.png">

Either the command call is missing something, or openshift-tools has an issue. @dgolovin which is it?

dgolovin replied:
@fbricon will look at this, it could be odo failure, bz project already has devfile.yaml. I am going to fix it by detecting project with devfile.yaml in root folder and add option `--devfile devfile.yaml`, so it will just use existing file.

dgolovin replied:
@fbricon https://github.com/redhat-developer/vscode-openshift-tools/issues/1815 is fixed. Released to master branch.

datho7561 replied:
@dgolovin Thanks for the update! I looked over the code, and it seems I was just calling the command with the wrong parameters. It should work now. I'll record a short demo of the new workflow.

dgolovin commented on src/utils/openShiftConnectorUtils.ts:
```diff
@@ -0,0 +1,77 @@
+      })
+    ]);
+  } catch (e) {
+    installDisposables.forEach((d: Disposable) => { d.dispose(); });
```
The code creates Promise instance and returns it. This catch block is basically dead branch IMO, it will never get control. Exceptions inside callbacks will be caught at line 50 in `deployToOpenShift.ts`.

datho7561 replied:
Good catch, thanks! I'll fix this later today

datho7561 commented on src/wizards/deployToOpenShift/deployToOpenShift.ts:
```diff
@@ -0,0 +1,61 @@
+  try {
+    await installOpenShiftConnectorIfNeeded();
+    const quarkusProject: ProjectLabelInfo = await getQuarkusProject();
+    const result: string = await commands.executeCommand("openshift.component.createFromRootWorkspaceFolder", //
```
I am looking into trying to push the component, then activating the "create component" if that fails. I am hoping this will cover the use case where there is a `devfile.yaml` and a `.odo` folder, but the component is missing from OpenShift. I'm also going to wrap the command calls to a different interface to make this function cleaner to read.

datho7561 replied:
I changed strategies to checking the folder for `.odo` to detect if the component should be created or just pushed.

datho7561 commented on src/wizards/deployToOpenShift/deployToOpenShift.ts:
```diff
@@ -0,0 +1,61 @@
+    }
+    return window.showInformationMessage(result);
+  } catch (error) {
+    return window.showErrorMessage(`Failed to deploy to OpenShift`);
```
I need to remove the error/success reporting here, since OpenShift connector will handle this

datho7561 committed:
Deploy to OpenShift command

Uses the existing menu to select which Quarkus project to deploy.
Uses OpenShift Connector in order to deploy the Quarkus project to
OpenShift.
The component is created if it does not exist yet (i.e. no .odo folder
is present), and it's pushed if it already exists.
OpenShift Connector handles selecting/creating the application and
naming the component.

Closes #313

Signed-off-by: David Thompson <davthomp@redhat.com>

dgolovin reviewed (approved):
Except the fact that some `return Promise.resolve();` and some try/catch blocks are not required,  It looks good to me. Haven't tried it though.

datho7561 commented on src/wizards/deployToOpenShift/deployToOpenShift.ts:
```diff
@@ -0,0 +1,65 @@
+import { Uri, window } from "vscode";
```
If you try to deploy a Quarkus project, but aren't connected to a cluster, then the error message is not friendly.

datho7561 replied:
I don't seem to get any error message, but the dialog to push doesn't show up if there is no cluster connected.

datho7561 replied:
If you create the component when the cluster is not connected, you do not get validation on the project field, so it could be easy to get the config wrong. It might make sense to only allow deploying to OpenShift through this command when vscode-openshift is able to connect to a cluster.

datho7561 replied:
I implemented a hack where it runs `kubectl api-resources` to see if the computer is connected to a cluster. If not connected, then it doesn't try to create the component.

datho7561 commented on src/utils/openShiftConnectorUtils.ts:
```diff
@@ -0,0 +1,128 @@
+    })
+  ]).catch(e => {
+    installListenerDisposables.forEach((d: Disposable) => { d.dispose(); });
+    return Promise.reject(e);
```
`throw e`

dgolovin commented on src/utils/openShiftConnectorUtils.ts:
```diff
@@ -0,0 +1,148 @@
+ * @throws if the user refuses to install the extension, or if the extension does not get installed within a timeout period
+ */
+async function installOpenShiftConnector(): Promise<void> {
+  const installListenerDisposables: Disposable[] = [];
```
Should be just `installListenerDisposable: Disposable;` because it is always one returned from `onDidChange` call. That will reduce `forEach` to just single call of `dispose()`.

dgolovin commented on src/utils/openShiftConnectorUtils.ts:
```diff
@@ -0,0 +1,148 @@
+      // Fail install if it takes longer than the timeout period
+      setTimeout(reject, DOWNLOAD_TIMEOUT, new Error(`${OPENSHIFT_CONNECTOR} installation is taking a while. Cancelling 'Deploy to OpenShift'`));
+    })
+  ]).catch(e => {
```
should be
```
]).finally(() => {
    installListenerDisposable.dispose();
})
```

dgolovin commented on src/utils/openShiftConnectorUtils.ts:
```diff
@@ -0,0 +1,148 @@
+  const installListenerDisposables: Disposable[] = [];
+  return Promise.race([
+    new Promise<void>((resolve, reject) => {
+      extensions.onDidChange(() => {
```
```installListenerDisposable = extensions.onDidChange(() => {```

dgolovin commented on src/utils/openShiftConnectorUtils.ts:
```diff
@@ -0,0 +1,148 @@
+    new Promise<void>((resolve, reject) => {
+      extensions.onDidChange(() => {
+        if (isOpenShiftConnectorInstalled()) {
+          installListenerDisposables.forEach((d: Disposable) => { d.dispose(); });
```
This line is not needed when using `finally()` see change below.

dgolovin commented on src/utils/openShiftConnectorUtils.ts:
```diff
@@ -0,0 +1,148 @@
+          installListenerDisposables.forEach((d: Disposable) => { d.dispose(); });
+          resolve();
+        }
+      }, null, installListenerDisposables);
```
These parameters are not needed.

dgolovin commented on src/utils/openShiftConnectorUtils.ts:
```diff
@@ -0,0 +1,148 @@
+    })
+  ]).catch(e => {
+    installListenerDisposables.forEach((d: Disposable) => { d.dispose(); });
+    throw e;
```
No need to re-throw the exception.

dgolovin commented on src/wizards/deployToOpenShift/deployToOpenShift.ts:
```diff
@@ -0,0 +1,66 @@
+        return getOpenShiftConnector();
+      }
+    } catch (e) {
+      throw new Error(`${OPENSHIFT_CONNECTOR} installation failed`);
```
I am running vscode-quarkus from dev env. Whenever I simulate this exception the error is only reported in console, it is never reported to the user through `window.showErrorMessage`.

dgolovin commented on src/utils/openShiftConnectorUtils.ts:
```diff
@@ -0,0 +1,150 @@
+ * @returns when the extension is installed
+ * @throws if the user refuses to install the extension, or if the extension does not get installed within a timeout period
+ */
+async function installOpenShiftConnector(): Promise<void> {
```
Even better version for this function.
Notice: 
1. no Promise.race() required anymore
2. no unused parameters in callback
3. Promise.resolve(commands.executeCommand()) call converts `thenable` instance to `Promise` and that in turn allows to use just `.catch(reject)`.
4. Even when timer expired after Promise fulfilled, rejecting it won't have any effect on fulfilled promise
5. ?. is not required because it `installListenerDisposable` is always going to be initialized
``` javascript
async function installOpenShiftConnector(): Promise<void> {
  let installListenerDisposable: Disposable;

  return new Promise<void>((resolve, reject) => {
    installListenerDisposable = extensions.onDidChange(() => {
      if (isOpenShiftConnectorInstalled()) {
        resolve();
      }
    });
    Promise.resolve(commands.executeCommand("workbench.extensions.installExtension", OPENSHIFT_CONNECTOR_EXTENSION_ID)).catch(reject);
    setTimeout(reject, DOWNLOAD_TIMEOUT, new Error(`${OPENSHIFT_CONNECTOR} installation is taking a while. Cancelling 'Deploy to OpenShift'. Please retry after the OpenShift Connector installation has finished`));
  }).finally(() => {
    installListenerDisposable.dispose();
  });
}
```

datho7561 commented on src/utils/openShiftConnectorUtils.ts:
```diff
@@ -0,0 +1,97 @@
+
+export const OPENSHIFT_CONNECTOR_EXTENSION_ID: string = 'redhat.vscode-openshift-connector';
+export const OPENSHIFT_CONNECTOR = 'OpenShift Connector extension';
+const DOWNLOAD_TIMEOUT: number = 60000; // Timeout for downloading VSCode OpenShift Connector, in milliseconds
```
Remove typing
